Fix check_expr_if to point to a more accurate location of the compilation error in some cases

sgasho opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/147484)*
<!-- homu-ignore:end -->

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

closes rust-lang/rust#146190

The problem linked on the issue above lays on how Expr::If is processed.

Currently, Expr::If is processed by depth-first like approach, which is natural considering that Expr::If is a nested structure.

e.g. 

```rust
fn main() {
    let oa = Some(1);
    let oa2 = Some(1);
    let v = if let Some(a) = oa {
        Some(&a)
    } else if let Some(a) = oa2 {
        &Some(a)
    } else {
        None
    };
    println!("{v:?}");
}
```

Simplified Expr::If structure for this code is like below.

```
Expr::If {
  cond: "if let Some(a) = oa",
  then: "Some(&a)",
  else: {
    Expr::If {
      cond: "else if let Some(a) = oa2",
      then: "&Some(a)",
      else: {
        cond: "else",
        then: "None",
        else: None,
      }
    }
  }
}
```

Current check_expr_if processes this Expr::If by depth-first-search like approach, that is why "`if` and `else` have incompatible types" occurs between else if let Some(a) = oa2 and else, not  between if let Some(a) = oa and if let Some(a) = oa2.

New approach on this PR is first flatten the Expr::If nested nodes, then process forward.

I created _if.rs and moved the check_expr_if becase this new approach requires a bit more lines of codes.

I added a test case that guarantees the problem on the linked issue has been resolved.
tests/ui/expr/if/if-multi-else-if-incompatible-types-issue-146190.rs


Discussion on Zulip: https://rust-lang.zulipchat.com/#narrow/channel/182449-t-compiler.2Fhelp/topic/Struggling.20to.20fix.20issue.20.23146190.20check_expr_if/with/540960074

fmease commented:
r? types

jackh726 commented on tests/ui/inference/deref-suggestion.rs:
Please revent the changes to this file, since it just seems like a drive-by change.

jackh726 commented on tests/ui/typeck/consider-borrowing-141810-3.stderr:
Same

sgasho replied:
removed this borrowing suggestions in this commit!

[fix: early returns suggest_deref_or_ref when the expr is an empty block](https://github.com/rust-lang/rust/pull/147484/commits/0fa922078f2adc09bdfb97d8d3982bc618f1fe07)

jackh726 reviewed (commented):
Perf looks good, so that's good to see.

Haven't yet reviewed the meat of this, but some comments. If you want to try to address them, that's great. If you want to wait until I review the meat of this, that's okay too.

I'll try to get back to this later today or this weekend, but want to share what I have.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147484/4f08307f6efdbca86d27beecd257fad9913192e9..5c38b514496e6b3b02079669f24ea07cd0431554)*

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,281 @@
+        }
+        self.diverges.set(initial_diverges | tail_diverges);
+    }
+}
```
I'm actually not a fan of splitting this (along with `reset_diverges_to_maybe` and `take_diverges`) into separate functions.

I guess, the more I look over this, I also am not a huge fan of `explain_if_branch_mismatch` either.

jackh726 reviewed (commented):
Sorry, another round of review. These changes I am going to pause for you to address before I get back to this.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147484/4f08307f6efdbca86d27beecd257fad9913192e9..5c38b514496e6b3b02079669f24ea07cd0431554)*

sgasho commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,281 @@
+            if let hir::ExprKind::If(..) = else_expr.kind {
+                current_if = else_expr;
+                continue;
+            }
```
removed duplicate logics and added some extra refactorings(sorry it if is hard to check diffs....)

[refactor: remove duplicate logic for checking Expr content. and some …](https://github.com/rust-lang/rust/pull/147484/commits/c52e5f9b4794de6a263b40c288b9f2be23a52ec7)

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,297 @@
+                .guarded_branches
+                .get(idx + 1)
+                .map(|next| next.expr_with_parts.expr)
+                .or(chain.last_expr().into());
```
Hmm, this doesn't seem right (or at least, is too subtle for my liking): when idx is len-1, `chain.last_expr()` is called and there is no final else, uses the *current* expr (since it is the last one).

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,297 @@
+                self.misc(branch_body.span)
+            };
+            let cause_span =
+                if idx == 0 { Some(root_if_expr.span) } else { Some(branch_body.span) };
```
If `idx == 0`, isn't branch_body.span == root_if_expr.span?

jackh726 commented on compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs:
```diff
@@ -628,19 +628,29 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                     .expect("if expression only expected inside FnCtxt")
-                    .expr_ty(else_expr);
-                if let hir::ExprKind::If(_cond, _then, None) = else_expr.kind
+                    .expr_ty_opt(else_expr);
```
Hmm, so...is this ever going to return `Some`?

jackh726 reviewed (commented):
Some more review comments. Looking better, but still have some questions/thoughts.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147484/733108b6d4acaa93fe26ae281ea305aacd6aac4e..fdd59f68c8be37fa34d2aaba4454facce2ea0fcd)*

sgasho committed:
fix: diagnostics - check if - else chain types from top to bottom, not from depth first approach

sgasho committed:
old approach (squashed for revert)

sgasho committed:
Revert "old approach (squashed for revert)"

This reverts commit ac49139fe0a60e54b19bd74196609ed64480614a.

rust-log-analyzer commented:
The job **`x86_64-gnu-miri`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25313851274/job/74206863909) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/74206863909) [(plain)](https://github.com/rust-lang/rust/commit/dddffd294742c038f5253f5864579d1163703080/checks/74206863909/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25314181326/job/74207933055) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/74207933055) [(plain)](https://github.com/rust-lang/rust/commit/1f3fe1e648239c36984ba372e4debcdcab463161/checks/74207933055/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] rustc_mir_transform test:false 5.525
error[E0061]: this method takes 1 argument but 2 arguments were supplied
   --> compiler/rustc_hir_typeck/src/_if.rs:132:27
    |
132 |             orig_expected.try_structurally_resolve_and_adjust_for_branches(self, if_span);
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^       ------- unexpected argument #2 of type `rustc_span::Span`
    |
note: method defined here
   --> compiler/rustc_hir_typeck/src/expectation.rs:43:19
    |
 43 |     pub(super) fn try_structurally_resolve_and_adjust_for_branches(
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: remove the extra argument
    |
132 -             orig_expected.try_structurally_resolve_and_adjust_for_branches(self, if_span);
132 +             orig_expected.try_structurally_resolve_and_adjust_for_branches(self);
    |

For more information about this error, try `rustc --explain E0061`.
[RUSTC-TIMING] rustc_hir_typeck test:false 6.414
error: could not compile `rustc_hir_typeck` (lib) due to 1 previous error
```

</details>

sgasho commented:
@jackh726 

Changed direction. The 7 old commits are squashed + reverted in history so both designs are reviewable.

Keeping the code architecture clean with the flattening approach turned out to be harder than I
expected, so I rethought it.

The new approach keeps the nested `Expr::If` and passes parent ty to children to coerce them.

Also, I added a test case for an if-chain nested inside another if-chain at tests/ui/typeck/if-else-type-mismatch-chain-nested.rs.

FWIW, I made some simple illustrations of the before/after approaches.

#### Before

<img width="1211" height="549" alt="スクリーンショット 2026-05-04 19 55 21" src="https://github.com/user-attachments/assets/e4551b4f-329f-4e6c-b0cb-0e258d5fe67d" />


#### After

<img width="470" height="406" alt="スクリーンショット 2026-05-04 20 17 35" src="https://github.com/user-attachments/assets/660ef1a5-2471-40c1-bdff-337bf42a9074" />

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-21-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25314665793/job/74209431645) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/74209431645) [(plain)](https://github.com/rust-lang/rust/commit/57eeb1986535e7378387d463951b40219bc4c585/checks/74209431645/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
Executing "/scripts/stage_2_test_set1.sh"
+ /scripts/stage_2_test_set1.sh
+ '[' 1 == 1 ']'
+ echo 'PR_CI_JOB set; skipping tidy'
+ SKIP_TIDY='--skip tidy'
+ ../x.py --stage 2 test --skip tidy --skip compiler --skip src
PR_CI_JOB set; skipping tidy
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
---
15 error: aborting due to 1 previous error
16 

Note: some mismatched output was normalized before being compared
- LL |     } else if false &{ //~ ERROR `if` and `else` have incompatible types [E0308]
+    |
+ help: consider borrowing here
+    |
+ LL |     } else if false &{
+    |                     +
---
To only update this specific test, also pass `--test-args typeck/consider-borrowing-141810-2.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/typeck/consider-borrowing-141810-2.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/typeck/consider-borrowing-141810-2" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "incomplete_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error[E0308]: `if` and `else` have incompatible types
##[error]  --> /checkout/tests/ui/typeck/consider-borrowing-141810-2.rs:4:21
   |
LL |        let x = if true {
   |  ______________-
LL | |          &()
   | |          --- expected because of this
LL | |      } else if false { //~ ERROR `if` and `else` have incompatible types [E0308]
   | | _____________________^
LL | ||     } else {
   | ||_____^ expected `&()`, found `()`
LL | |      };
   | |______- `if` and `else` have incompatible types
   |
help: consider borrowing here
   |
LL |     } else if false &{ //~ ERROR `if` and `else` have incompatible types [E0308]
   |                     +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
```

</details>

sgasho commented:
@rustbot ready

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,232 @@
+        &self,
+        cx: &mut ElseIfChainCx<'_, 'tcx>,
+        prev_branch: PrevBranch<'tcx>,
+        current_expr: &'tcx hir::Expr<'tcx>,
```
This should like take the `cond_expr`, `then_expr` and `opt_else_expr` directly.

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,232 @@
+        coerce.coerce(self, &self.misc(sp), then_expr, then_ty);
+
+        if let Some(else_expr) = opt_else_expr {
+            let else_diverges = if chain_has_terminal_else_block(else_expr)
```
What is the `chain_has_terminal_else_block` check for? What's wrong with copying the `None` case to `check_else_if_branch`?

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,232 @@
+                //     } else if c2 &{
+                // which isn't valid syntax. Having an empty block in an if chain is an edge
+                // case in real code, so the fallback is harmless.
+                && !has_empty_block_chains(else_expr)
```
This feels...potentially insufficient? Is it not true that we could get the same suggestion with an explicit `()`? Or are we pointing to the block *because* there isn't an explicit type?

In either case, it seems like it would be better to fix that elsewhere? And to keep the logic here simple.

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,232 @@
+
+        coerce.coerce(self, &self.misc(sp), then_expr, then_ty);
+
+        if let Some(else_expr) = opt_else_expr {
```
`check_else_if_branch` uses a match, which...I like better than the nested ifs here.

jackh726 commented on compiler/rustc_hir_typeck/src/_if.rs:
```diff
@@ -0,0 +1,232 @@
+impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+    // A generic function for checking the 'then' and 'else' clauses in an 'if'
+    // or 'if-else' expression.
+    pub(crate) fn check_expr_if(
```
So, `check_expr_if` and `check_else_if_branch` are *very* similar, and I wonder if they could be combined - or at least with considerably more code combined.

jackh726 reviewed (commented):
Apologies for the long-awaited review here. It's tricky code and I wanted to be sure to give it a proper review.

I think this setup looks fairly decent, but I think more code can be shared to reduce the added complexity.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147484/1d72d7e8136faaebad3a85eeed432e6ea1b2ffab..7d6060c17895b70d9b663c1a57fa175ec9a5d259)*
